daemon/api_sideload_n_try.go

```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
+	}
+	if errRsp != nil {
+		return errRsp
+	}
+
+	systemRestartImmediate := isTrue(form, "system-restart-immediate")
+	if systemRestartImmediate {
+		chg.Set("system-restart-immediate", true)
+	}
+
+	ensureStateSoon(st)
+
+	// the handoff is only done when the unlock succeeds (instead of panicking)
+	// but this is good enough
+	pathsToNotRemove = tempPaths
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func sideloadManySnaps(st *state.State, form *Form, origPaths, tempPaths []string, flags snapstate.Flags, user *auth.UserState) (*state.Change, *apiError) {
+	var sideInfos []*snap.SideInfo
+	var names []string
+
+	for i, tempPath := range tempPaths {
+		si, apiError := readSideInfo(st, form, tempPath, origPaths[i])
+		if apiError != nil {
+			return nil, apiError
+		}
+
+		sideInfos = append(sideInfos, si)
+		names = append(names, si.RealName)
+	}
+
+	var userID int
+	if user != nil {
+		userID = user.ID
+	}
+
+	tss, err := snapstateInstallPathMany(context.TODO(), st, sideInfos, tempPaths, userID, &flags)
+	if err != nil {
+		return nil, errToResponse(err, tempPaths, InternalError, "cannot install snap files: %v")
+	}
+
+	msg := fmt.Sprintf(i18n.G("Install snaps %q from files %q"), names, origPaths)
+	chg := newChange(st, "install-snaps", msg, tss, names)
+	chg.Set("api-data", map[string][]string{"snap-names": names})
+
+	return chg, nil
+}
+
+func sideloadSnap(st *state.State, form *Form, origPath, tempPath string, flags snapstate.Flags) (*state.Change, *apiError) {
 	var instanceName string
 	if len(form.Values["name"]) > 0 {
 		// caller has specified desired instance name
 		instanceName = form.Values["name"][0]
 		if err := snap.ValidateInstanceName(instanceName); err != nil {
-			return BadRequest(err.Error())
+			return nil, BadRequest(err.Error())
 		}
 	}
 
-	st := c.d.overlord.State()
-	st.Lock()
-	defer st.Unlock()
+	sideInfo, apiErr := readSideInfo(st, form, tempPath, origPath)
+	if apiErr != nil {
+		return nil, apiErr
+	}
 
-	var snapName string
+	if instanceName != "" {
+		requestedSnapName := snap.InstanceSnap(instanceName)
+		if requestedSnapName != sideInfo.RealName {
+			return nil, BadRequest(fmt.Sprintf("instance name %q does not match snap name %q", instanceName, sideInfo.RealName))
+		}
+	} else {
+		instanceName = sideInfo.RealName
+	}
+
+	tset, _, err := snapstateInstallPath(st, sideInfo, tempPath, instanceName, "", flags)
+	if err != nil {
+		return nil, errToResponse(err, []string{sideInfo.RealName}, InternalError, "cannot install snap file: %v")
+	}
+
+	msg := fmt.Sprintf(i18n.G("Install %q snap from file %q"), instanceName, origPath)
+	chg := newChange(st, "install-snap", msg, []*state.TaskSet{tset}, []string{instanceName})
+	chg.Set("api-data", map[string]string{"snap-name": instanceName})
+
+	return chg, nil
+}
+
+func readSideInfo(st *state.State, form *Form, tempPath string, origPath string) (*snap.SideInfo, *apiError) {
 	var sideInfo *snap.SideInfo
 
+	dangerousOK := isTrue(form, "dangerous")
 	if !dangerousOK {
```

Comment:
```suggestion
	if !flags.DangerousOK {
```

and the `!isTrue(form, "devmode")` further down can be replaced with `flags.SnapModeFlags & snapstate.DevMode == 0` (or could be a sideloadFlags level thing itself